onnx/external_data_helper.py

```diff
@@ -255,11 +255,9 @@ def _get_attribute_tensors(onnx_model_proto: ModelProto) -> Iterable[TensorProto
 
 
 def _sanitize_path(path: str) -> str:
-    """Remove path components which would allow traversing up a directory tree from a base path.
-
-    Note: This method is currently very basic and should be expanded.
-    """
-    return path.lstrip("/.")
+    """Remove path components which would allow traversing up a directory tree from a base path."""
+    path = os.path.normpath(path)
```

Comment:
How does it fix a reference upward? As I read the documentation, it only removes redundant ".." ... what happens for non-redundant ".."?

Reply:
What does onnxruntime do for external tensor name sanitization?

Reply:
will os.path.basename(path) avoid 'path' being pointed to location outside of cwd? Here is my experiment:
>>> path="/external_data/../../secret"
>>> import os
>>> path2 = os.path.normpath(path)
>>> path2
'\\secret'
>>> path3 = os.path.basename(path)
>>> path3
'secret'

Reply:
update to use lstrip to avoid traversing. normpath will put all "\/." at the front. remove them so that final path is within base dir.

Reply:
I see that this is what we do in the C++ side: https://github.com/onnx/onnx/blob/b833ee7d1675ea183a12706670b0006340a40040/onnx/checker.cc#L131 ... I wonder if we should convert that into a C++ utility, and call it from python? So that the logic is one place?

Reply:
https://security.openstack.org/guidelines/dg_using-file-paths.html#correct 

is this helpful?

Reply:
Also 

```
   # Check that the file exists at the path  
   if not os.path.isfile(image_path):
       abort(404)

   # Get the real, absolute path to the file
   safe_path = os.path.realpath(image_path)

   common_base = os.path.commonpath([base_path, safe_path]) 
   if common_base != base_path:
        abort(403)

   # Verify basename of resolved path matches original file
    if os.path.basename(safe_path) != image_name:
        # Invalid - path traversal detected
        abort(403) 
```

from https://learn.snyk.io/lesson/directory-traversal/

Reply:
> What does onnxruntime do for external tensor name sanitization?

ORT code is in c++. (https://github.com/microsoft/onnxruntime/blob/1007d8f3d1904ff2efc4e0647e795939fe049464/onnxruntime/core/framework/tensorprotoutils.cc#L148)
it simply concatenate base dir with location filename.

Reply:
So we use the `base_path` in https://github.com/onnx/onnx/blob/29c230d9eee40e704a6a2c437ff2141df6821a6f/onnx/external_data_helper.py#L32C56-L32C64 for the validation

Reply:
> https://security.openstack.org/guidelines/dg_using-file-paths.html#correct
> 
> is this helpful?

it does help. I was also thinking to extend _sanitize_path to take base_dir which makes more sense.

Reply:
> I see that this is what we do in the C++ side:
> 
> https://github.com/onnx/onnx/blob/b833ee7d1675ea183a12706670b0006340a40040/onnx/checker.cc#L131
> 
> ... I wonder if we should convert that into a C++ utility, and call it from python? So that the logic is one place?

given that both ORT and ONNX need to properly sanitize external data path, lets do it in c++. I will update this PR.